You have to work on the files:
*  [Books](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Books.csv.gz)
*  [Book ratings](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Book-Ratings.csv.gz)
*  [Users](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Users.csv.gz)
*  [Goodbooks books](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks.csv.gz)
*  [Goodbooks ratings](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks-ratings.csv.gz)

### Notes

1.    It is mandatory to use GitHub for developing the project.
1.    The project must be a jupyter notebook.
1.    There is no restriction on the libraries that can be used, nor on the Python version.
1.    To read those files, you need to use the `encoding = 'latin-1'` option.
1.    All questions on the project **must** be asked in a public channel on [Zulip](https://focs.zulipchat.com), otherwise no  answer will be given.

In [2]:
import pandas as pd
import numpy as np

In [3]:
books = pd.read_csv("/Users/andreaermellino/desktop/fcs/Books.csv", encoding = 'latin-1', sep=';', low_memory = False)
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
goodbooks = pd.read_csv("/Users/andreaermellino/desktop/fcs/goodbooks.csv", encoding = 'latin-1', sep = ',')
goodbooks.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0        1            2767052       2767052  2792775          272  439023483   
1        2                  3             3  4640799          491  439554934   
2        3              41865         41865  3212258          226  316015849   
3        4               2657          2657  3275794          487   61120081   
4        5               4671          4671   245494         1356  743273567   

         isbn13                       authors  original_publication_year  \
0  9.780439e+12               Suzanne Collins                     2008.0   
1  9.780440e+12  J.K. Rowling, Mary GrandPré                     1997.0   
2  9.780316e+12               Stephenie Meyer                     2005.0   
3  9.780061e+12                    Harper Lee                     1960.0   
4  9.780743e+12           F. Scott Fitzgerald                     1925.0   

                             original_title  ... ratings_count  \
0                          The Hunger Games  ...       4780653   
1  Harry Potter and the Philosopher's Stone  ...       4602479   
2                                  Twilight  ...       3866839   
3                     To Kill a Mockingbird  ...       3198671   
4                          The Great Gatsby  ...       2683664   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   
1            4800065                    75867      75504     101676   
2            3916824                    95009     456191     436802   
3            3340896                    72586      60427     117415   
4            2773745                    51992      86236     197621   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   
1     455024    1156318    3011543   
2     793319     875073    1355439   
3     446835    1001952    1714267   
4     606158     936012     947718   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1474154022m...   
2  https://images.gr-assets.com/books/1361039443m...   
3  https://images.gr-assets.com/books/1361975680m...   
4  https://images.gr-assets.com/books/1490528560m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1474154022s...  
2  https://images.gr-assets.com/books/1361039443s...  
3  https://images.gr-assets.com/books/1361975680s...  
4  https://images.gr-assets.com/books/1490528560s...  

[5 rows x 23 columns]

In [5]:
gb_ratings = pd.read_csv("/Users/andreaermellino/desktop/fcs/goodbooks-ratings.csv", encoding = 'latin-1', sep = ',')
gb_ratings.head()

user_id  book_id  rating
0        1      258       5
1        2     4081       4
2        2      260       5
3        2     9296       5
4        2     2318       3

In [56]:
users = pd.read_csv("/Users/andreaermellino/desktop/fcs/Users.csv", encoding = 'latin-1', sep = ';')
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [7]:
book_ratings = pd.read_csv("/Users/andreaermellino/desktop/fcs/book-ratings.csv", encoding = 'latin-1', sep = ';')
book_ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

### 1. Normalize the location field of *Users* dataset, splitting into city, region, country.

Due strade alternative.

1)

In [58]:
users[['city','state','country']] = users['Location'].str.split(',', expand=True, n=2)
del(users['Location'])
users.head()

User-ID   Age         city             state          country
0        1   NaN          nyc          new york              usa
1        2  18.0     stockton        california              usa
2        3   NaN       moscow   yukon territory           russia
3        4  17.0        porto          v.n.gaia         portugal
4        5   NaN  farnborough             hants   united kingdom

2)

In [15]:
location = users['Location']

In [16]:
city = location.str.extract(r'(^([a-z]*[\s]*)*)')
city = city[0]
city.head()
users['City'] = city
users.head()

User-ID                            Location   Age         City
0        1                  nyc, new york, usa   NaN          nyc
1        2           stockton, california, usa  18.0     stockton
2        3     moscow, yukon territory, russia   NaN       moscow
3        4           porto, v.n.gaia, portugal  17.0        porto
4        5  farnborough, hants, united kingdom   NaN  farnborough

In [17]:
region = location.str.extract(r'(, .*,)') # r'(, ([a-z]*[\s]*)*,)'
region = region[0]
region = region.str.replace(', ', '')
region = region.str.replace(',', '')
region.head()
users['Region'] = region
users.head()

User-ID                            Location   Age         City  \
0        1                  nyc, new york, usa   NaN          nyc   
1        2           stockton, california, usa  18.0     stockton   
2        3     moscow, yukon territory, russia   NaN       moscow   
3        4           porto, v.n.gaia, portugal  17.0        porto   
4        5  farnborough, hants, united kingdom   NaN  farnborough   

            Region  
0         new york  
1       california  
2  yukon territory  
3         v.n.gaia  
4            hants

In [18]:
country = location.str.extract(r'(([a-zA-Z]+( [a-zA-Z]*)*)$)')
country = country[0]

In [19]:
country.head()
users['Country'] = country
users.head()
users.drop(columns = ['Location'], inplace = True)
users.head()

User-ID   Age         City           Region         Country
0        1   NaN          nyc         new york             usa
1        2  18.0     stockton       california             usa
2        3   NaN       moscow  yukon territory          russia
3        4  17.0        porto         v.n.gaia        portugal
4        5   NaN  farnborough            hants  united kingdom

### 2. For each book in the *Books* dataset, compute its average rating.

In [20]:
media_rating = pd.merge(books, book_ratings, how = 'left', on = 'ISBN')
media_rating.head()

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   
2  0002005018         Clara Callan  Richard Bruce Wright                2001   
3  0002005018         Clara Callan  Richard Bruce Wright                2001   
4  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
3    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
4    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0002005018.0...   
3  http://images.amazon.com/images/P/0002005018.0...   
4  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
0  http://images.amazon.com/images/P/0195153448.0...      2.0          0.0  
1  http://images.amazon.com/images/P/0002005018.0...      8.0          5.0  
2  http://images.amazon.com/images/P/0002005018.0...  11400.0          0.0  
3  http://images.amazon.com/images/P/0002005018.0...  11676.0          8.0  
4  http://images.amazon.com/images/P/0002005018.0...  41385.0          0.0

In [21]:
media = media_rating.groupby('ISBN').mean()['Book-Rating']

In [22]:
book_rating_media = pd.merge(books, media, on='ISBN')[['ISBN', 'Book-Title', 'Book-Author', 'Book-Rating']]
book_rating_media.columns = ['ISBN', 'Book-Title', 'Book-Author', 'AVG Rating']
book_rating_media.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author  AVG Rating  
0    Mark P. O. Morford    0.000000  
1  Richard Bruce Wright    4.928571  
2          Carlo D'Este    5.000000  
3      Gina Bari Kolata    4.272727  
4       E. J. W. Barber    0.000000

### 3. For each book in the *GoodBooks* dataset, compute its average rating.

In [17]:
goodbooks.shape

(99, 23)

In [18]:
gb_ratings.shape

(99, 3)

In [19]:
goodbooks.head(10)

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
0        1            2767052       2767052   2792775          272   
1        2                  3             3   4640799          491   
2        3              41865         41865   3212258          226   
3        4               2657          2657   3275794          487   
4        5               4671          4671    245494         1356   
5        6           11870085      11870085  16827462          226   
6        7               5907          5907   1540236          969   
7        8               5107          5107   3036731          360   
8        9                960           960   3338963          311   
9       10               1885          1885   3060926         3455   

         isbn        isbn13                       authors  \
0   439023483  9.780439e+12               Suzanne Collins   
1   439554934  9.780440e+12  J.K. Rowling, Mary GrandPré   
2   316015849  9.780316e+12               Stephenie Meyer   
3    61120081  9.780061e+12                    Harper Lee   
4   743273567  9.780743e+12           F. Scott Fitzgerald   
5   525478817  9.780525e+12                    John Green   
6   618260307  9.780618e+12                J.R.R. Tolkien   
7   316769177  9.780317e+12                 J.D. Salinger   
8  1416524797  9.781417e+12                     Dan Brown   
9   679783261  9.780680e+12                   Jane Austen   

   original_publication_year                            original_title  ...  \
0                     2008.0                          The Hunger Games  ...   
1                     1997.0  Harry Potter and the Philosopher's Stone  ...   
2                     2005.0                                  Twilight  ...   
3                     1960.0                     To Kill a Mockingbird  ...   
4                     1925.0                          The Great Gatsby  ...   
5                     2012.0                    The Fault in Our Stars  ...   
6                     1937.0        The Hobbit or There and Back Again  ...   
7                     1951.0                    The Catcher in the Rye  ...   
8                     2000.0                          Angels & Demons   ...   
9                     1813.0                       Pride and Prejudice  ...   

  ratings_count work_ratings_count  work_text_reviews_count  ratings_1  \
0       4780653            4942365                   155254      66715   
1       4602479            4800065                    75867      75504   
2       3866839            3916824                    95009     456191   
3       3198671            3340896                    72586      60427   
4       2683664            2773745                    51992      86236   
5       2346404            2478609                   140739      47994   
6       2071616            2196809                    37653      46023   
7       2044241            2120637                    44920     109383   
8       2001311            2078754                    25112      77841   
9       2035490            2191465                    49152      54700   

   ratings_2  ratings_3  ratings_4  ratings_5  \
0     127936     560092    1481305    2706317   
1     101676     455024    1156318    3011543   
2     436802     793319     875073    1355439   
3     117415     446835    1001952    1714267   
4     197621     606158     936012     947718   
5      92723     327550     698471    1311871   
6      76784     288649     665635    1119718   
7     185520     455042     661516     709176   
8     145740     458429     716569     680175   
9      86485     284852     609755    1155673   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1474154022m...   
2  https://images.gr-assets.com/books/1361039443m...   
3  https://images.gr-assets.com/books/1361975680m...   
4  https://images.gr-assets.com/books/1490528560m...   

In [23]:
def gb_mean(row):
    media = (row['ratings_1']*1 + row['ratings_2']*2 + row['ratings_3']*3 + row['ratings_4']*4 
             + row['ratings_5']*5)/row['work_ratings_count']
    return media

In [24]:
gb_media = goodbooks
gb_media['AVG Rating'] = goodbooks.apply(gb_mean, axis=1)
gb_media = gb_media[['book_id', 'authors', 'original_title', 'AVG Rating', 'isbn']]
gb_media.columns = ['BookID', 'Authors', 'Original Title', 'AVG Rating', 'ISBN']
gb_media.head()

BookID                       Authors  \
0       1               Suzanne Collins   
1       2  J.K. Rowling, Mary GrandPré   
2       3               Stephenie Meyer   
3       4                    Harper Lee   
4       5           F. Scott Fitzgerald   

                             Original Title  AVG Rating       ISBN  
0                          The Hunger Games    4.341984  439023483  
1  Harry Potter and the Philosopher's Stone    4.443047  439554934  
2                                  Twilight    3.571067  316015849  
3                     To Kill a Mockingbird    4.254818   61120081  
4                          The Great Gatsby    3.887376  743273567

### 4. Merge together all rows sharing the same book title, author and publisher. We will call the resulting datset `merged books`. The books that have not been merged together will not appear in `merged books`.

In [25]:
merged = books.groupby(['Book-Title', 'Book-Author', 'Publisher'], as_index=False).count()
merged = merged[merged['ISBN'] >= 2]
merged_books = merged[['Book-Title', 'Book-Author', 'Publisher']]

In [26]:
merged_books.head()

Book-Title  \
41                           !%@ (A Nutshell handbook)   
59   'A Hell of a Place to Lose a Cow': An American...   
255  10,000 dreams interpreted: A dictionary of dreams   
465                                   101 Famous Poems   
713                 15 Houseplants Even You Can't Kill   

                 Book-Author                       Publisher  
41             Donnalyn Frey                        O'Reilly  
59               Tim Brookes             National Geographic  
255  Gustavus Hindman Miller       Barnes &amp; Nobles Books  
465              Roy J. Cook  McGraw-Hill/Contemporary Books  
713                Joe Elder               Berkley Pub Group

In [21]:
books.shape

(271359, 8)

In [22]:
merged_books.shape

(4725, 3)

### 5. For each book in `merged books` compute its average rating.

The average is computed considering all books in `books` that have been merged.

In [27]:
merged_books_avg = pd.merge(books, merged_books, on=['Book-Title', 'Book-Author', 'Publisher'], how='right')[['ISBN', 'Book-Title', 'Book-Author', 'Publisher']]
merged_books_avg.head()

ISBN                                         Book-Title  \
0  1565920465                          !%@ (A Nutshell handbook)   
1  1565920317                          !%@ (A Nutshell handbook)   
2  0792276833  'A Hell of a Place to Lose a Cow': An American...   
3  0792277295  'A Hell of a Place to Lose a Cow': An American...   
4  1566196256  10,000 dreams interpreted: A dictionary of dreams   

               Book-Author                  Publisher  
0            Donnalyn Frey                   O'Reilly  
1            Donnalyn Frey                   O'Reilly  
2              Tim Brookes        National Geographic  
3              Tim Brookes        National Geographic  
4  Gustavus Hindman Miller  Barnes &amp; Nobles Books

In [28]:
book_ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [29]:
merged_books_ratings = pd.merge(merged_books_avg, book_ratings, how='left', on='ISBN')
merged_books_ratings.head()

ISBN                                         Book-Title  \
0  1565920465                          !%@ (A Nutshell handbook)   
1  1565920317                          !%@ (A Nutshell handbook)   
2  0792276833  'A Hell of a Place to Lose a Cow': An American...   
3  0792276833  'A Hell of a Place to Lose a Cow': An American...   
4  0792276833  'A Hell of a Place to Lose a Cow': An American...   

     Book-Author            Publisher   User-ID  Book-Rating  
0  Donnalyn Frey             O'Reilly  271245.0          0.0  
1  Donnalyn Frey             O'Reilly   87141.0          6.0  
2    Tim Brookes  National Geographic   79186.0          0.0  
3    Tim Brookes  National Geographic  114216.0          0.0  
4    Tim Brookes  National Geographic  201058.0          0.0

In [30]:
media = merged_books_ratings.groupby(['Book-Title', 'Book-Author', 'Publisher']).mean()['Book-Rating']
merged_books_avg = pd.merge(merged_books, media, on=['Book-Title', 'Book-Author', 'Publisher'])
merged_books_avg.columns = ['Book-Title', 'Book-Author', 'Publisher', 'AVG Rating']
merged_books_avg.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
2  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   
3                                   101 Famous Poems              Roy J. Cook   
4                 15 Houseplants Even You Can't Kill                Joe Elder   

                        Publisher  AVG Rating  
0                        O'Reilly    3.000000  
1             National Geographic    1.888889  
2       Barnes &amp; Nobles Books    7.000000  
3  McGraw-Hill/Contemporary Books    3.833333  
4               Berkley Pub Group    0.000000

### 6. For each book in `merged books` compute the minimum and maximum of the average ratings over all corresponding books in the `books` dataset.

Hence for each book in `merged books` we will have exactly two values (a minimum and a maximum)

In [31]:
merged_min_max = merged_books_ratings.groupby(['Book-Title','Book-Author','Publisher'], as_index=False)['Book-Rating'].agg([min,max])
merged_min_max.head()

min  \
Book-Title                                         Book-Author             Publisher                             
!%@ (A Nutshell handbook)                          Donnalyn Frey           O'Reilly                        0.0   
'A Hell of a Place to Lose a Cow': An American ... Tim Brookes             National Geographic             0.0   
10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller Barnes &amp; Nobles Books       4.0   
101 Famous Poems                                   Roy J. Cook             McGraw-Hill/Contemporary Books  0.0   
15 Houseplants Even You Can't Kill                 Joe Elder               Berkley Pub Group               0.0   

                                                                                                            max  
Book-Title                                         Book-Author             Publisher                             
!%@ (A Nutshell handbook)                          Donnalyn Frey           O'Reilly                         6.0  
'A Hell of a Place to Lose a Cow': An American ... Tim Brookes             National Geographic              9.0  
10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller Barnes &amp; Nobles Books       10.0  
101 Famous Poems                                   Roy J. Cook             McGraw-Hill/Contemporary Books  10.0  
15 Houseplants Even You Can't Kill                 Joe Elder               Berkley Pub Group                0.0

### 7. For each book in `goodbooks`, compute the list of its authors. Assuming that the number of reviews with a text (column `work_text_reviews_count`) is split equally among all authors, find for each authors the total number of reviews with a text. We will call this quantity the *shared number of reviews with a text*.

In [32]:
def find_author(lista):
    final = list()
    for authors in lista:
        single_book_list = authors.split(', ')
        for author in single_book_list:
            final.append(author)
    return final

In [33]:
def count_book_shared_number(row):
    shared_number = 0
    authors = row['authors'].split(', ')
    shared_number = row['work_text_reviews_count']/len(authors)
    return shared_number

In [34]:
gb_shared = goodbooks[['isbn', 'authors', 'original_title', 'work_text_reviews_count', 'original_publication_year']]
gb_shared['shared_number'] = gb_shared.apply(count_book_shared_number, axis=1)
gb_shared.head()

<ipython-input-34-c1a33f81cc58>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  gb_shared['shared_number'] = gb_shared.apply(count_book_shared_number, axis=1)


isbn                       authors  \
0  439023483               Suzanne Collins   
1  439554934  J.K. Rowling, Mary GrandPré   
2  316015849               Stephenie Meyer   
3   61120081                    Harper Lee   
4  743273567           F. Scott Fitzgerald   

                             original_title  work_text_reviews_count  \
0                          The Hunger Games                   155254   
1  Harry Potter and the Philosopher's Stone                    75867   
2                                  Twilight                    95009   
3                     To Kill a Mockingbird                    72586   
4                          The Great Gatsby                    51992   

   original_publication_year  shared_number  
0                     2008.0       155254.0  
1                     1997.0        37933.5  
2                     2005.0        95009.0  
3                     1960.0        72586.0  
4                     1925.0        51992.0

In [35]:
list_authors = find_author(goodbooks['authors'])
authors = pd.DataFrame(list_authors).drop_duplicates()
authors.columns = ['Name']

In [36]:
def count_author_shared_number(row):
    books_by_author = gb_shared[gb_shared['authors'].str.contains(row['Name'])]
    list_shared_number = list(books_by_author['shared_number'])
    shared_number = 0
    for i in range(0, len(list_shared_number)):
        shared_number += list_shared_number[i]
    return shared_number

In [37]:
complete_shared_number = authors.apply(count_author_shared_number, axis=1)
authors['Shared Number'] = complete_shared_number
authors.head()

Name  Shared Number
0  Suzanne Collins       340066.0
1     J.K. Rowling       136668.0
2   Mary GrandPré       136668.0
3  Stephenie Meyer       258573.0
4       Harper Lee        72586.0

### 8. For each year of publication, determine the author that has the largest value of the shared number of reviews with a text.

In [38]:
sn_per_year = gb_shared
sn_per_year['authors'] = sn_per_year['authors'].str.split(',')

<ipython-input-38-fb2aae7612b6>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sn_per_year['authors'] = sn_per_year['authors'].str.split(',')


In [39]:
sn_per_year.head(5)

isbn                          authors  \
0  439023483                [Suzanne Collins]   
1  439554934  [J.K. Rowling,  Mary GrandPré]   
2  316015849                [Stephenie Meyer]   
3   61120081                     [Harper Lee]   
4  743273567            [F. Scott Fitzgerald]   

                             original_title  work_text_reviews_count  \
0                          The Hunger Games                   155254   
1  Harry Potter and the Philosopher's Stone                    75867   
2                                  Twilight                    95009   
3                     To Kill a Mockingbird                    72586   
4                          The Great Gatsby                    51992   

   original_publication_year  shared_number  
0                     2008.0       155254.0  
1                     1997.0        37933.5  
2                     2005.0        95009.0  
3                     1960.0        72586.0  
4                     1925.0        51992.0

In [40]:
sn_per_year = sn_per_year.explode('authors')
sn_per_year.head()

isbn          authors                            original_title  \
0  439023483  Suzanne Collins                          The Hunger Games   
1  439554934     J.K. Rowling  Harry Potter and the Philosopher's Stone   
1  439554934   Mary GrandPré  Harry Potter and the Philosopher's Stone   
2  316015849  Stephenie Meyer                                  Twilight   
3   61120081       Harper Lee                     To Kill a Mockingbird   

   work_text_reviews_count  original_publication_year  shared_number  
0                   155254                     2008.0       155254.0  
1                    75867                     1997.0        37933.5  
1                    75867                     1997.0        37933.5  
2                    95009                     2005.0        95009.0  
3                    72586                     1960.0        72586.0

In [41]:
sn_per_year = sn_per_year.groupby(['original_publication_year', 'authors'], as_index=False).sum('shared_number')

In [42]:
best_sn_per_year = sn_per_year.loc[sn_per_year.groupby('original_publication_year')['shared_number'].idxmax()][['original_publication_year', 'authors', 'shared_number']]
best_sn_per_year.tail()

original_publication_year          authors  shared_number
130                     2009.0  Suzanne Collins        88538.0
131                     2010.0  Suzanne Collins        96274.0
132                     2011.0       E.L. James       103489.0
136                     2012.0       John Green       140739.0
138                     2015.0    Paula Hawkins        93600.0

### 9. Assuming that there are no errors in the ISBN fields, find the books in both datasets, and compute the difference of average rating according to the ratings and the goodratings datasets

In [43]:
ratings_merged = pd.merge(book_rating_media, gb_media, on='ISBN', how='inner')

In [44]:
ratings_merged['Difference'] = abs(ratings_merged['AVG Rating_x'] - ratings_merged['AVG Rating_y'])
ratings_merged.head()

ISBN                                         Book-Title  \
0  1400032717  The Curious Incident of the Dog in the Night-T...   
1  014028009X                              Bridget Jones's Diary   
2  014038572X                      The Outsiders (Now in Speak!)   
3  1594480001                                    The Kite Runner   
4  043965548X  Harry Potter and the Prisoner of Azkaban (Harr...   

       Book-Author  AVG Rating_x  BookID  \
0      Mark Haddon      4.813187      60   
1   Helen Fielding      3.751852      75   
2     S. E. Hinton      4.460674      90   
3  Khaled Hosseini      2.400000      11   
4     J.K. Rowling      3.533333      18   

                                    Authors  \
0                               Mark Haddon   
1                            Helen Fielding   
2                               S.E. Hinton   
3                           Khaled Hosseini   
4  J.K. Rowling, Mary GrandPré, Rufus Beck   

                                      Original Title  AVG Rating_y  Difference  
0  The Curious Incident of the Dog in the Night-Time      3.852457    0.960729  
1                                                NaN      3.746498    0.005354  
2                                      The Outsiders      4.060180    0.400494  
3                                   The Kite Runner       4.255954    1.855954  
4           Harry Potter and the Prisoner of Azkaban      4.527866    0.994533

In [45]:
ratings_merged = ratings_merged[['ISBN', 'Book-Title', 'Difference']]
ratings_merged.head()

ISBN                                         Book-Title  Difference
0  1400032717  The Curious Incident of the Dog in the Night-T...    0.960729
1  014028009X                              Bridget Jones's Diary    0.005354
2  014038572X                      The Outsiders (Now in Speak!)    0.400494
3  1594480001                                    The Kite Runner    1.855954
4  043965548X  Harry Potter and the Prisoner of Azkaban (Harr...    0.994533

### 10. Split the users dataset according to the age. One dataset contains the users with unknown age, one with age 0-14, one with age 15-24, one with age 25-34, and so on.

In [181]:
users['Age'].max()

244.0

Il valore più alto di età è pari a 244, ciò vuol dire che vogliamo dividere il dataset users in 25 subset: il primo contenente le righe di utenti con età sconosciuta, il secondo con le righe di utenti con età in 0-14 e poi proseguire con gruppi di età di 9 anni in 9 anni (15-24, 25-34 ... 235-244).
Siccome le uniche categorie 'anomale' sono quelle in cui l'età è sconosciuta o è nel range 0-14, l'estrazione di queste viene considerata come 'caso base' mentre l'estrazione delle righe appartenenti alle categorie 'ordinarie', ovvero quelle che vanno di 9 anni in 9 anni, viene fatta automaticamente e i risultati vengono memorizzati in un dizionario. Per accedere ai singoli dataset contenuti nel dizionario basta indicare l'etichetta corrispondente all'intervallo inferiore del range di età:

ESEMPIO: si vuole accedere al dataset in cui ci sono le righe con età in 155-164 -> age_dict['155']

In [98]:
age_null = pd.DataFrame()
age_0_14 = pd.DataFrame()
age_dict = {}
for x in range(15, int(users['Age'].max()), 10):
        age_dict[str(x)] = pd.DataFrame()

Siccome in users ci sono circa 279mila righe, i cicli for innestati richiedono un po' di tempo per produrre il risultato

In [102]:
for index, row in users.iterrows():
    if pd.isnull(row['Age']):
        age_null = age_null.append(row)
    elif row['Age'] >= 0 and row['Age'] <= 14:
        age_0_14 = age_0_14.append(row)
    elif row['Age'] > 14:
        for j in range(15, int(users['Age'].max()), 10):
            if row['Age'] >= j and row['Age'] <= j+9:
                age_dict[str(j)] = age_dict[str(j)].append(row)
                break

In [96]:
len(users)

278858

In [106]:
age_null.head()

Age         City         Country           Region  User-ID
0  NaN          nyc             usa         new york      1.0
2  NaN       moscow          russia  yukon territory      3.0
4  NaN  farnborough  united kingdom            hants      5.0
6  NaN   washington             usa               dc      7.0
7  NaN      timmins          canada          ontario      8.0

In [107]:
age_0_14.head()

Age          City    Country      Region  User-ID
10   14.0     melbourne  australia    victoria     11.0
18   14.0        weston        NaN                 19.0
40   14.0        santee        usa  california     41.0
119  13.0  villeneuve d     france        nord    120.0
219   0.0        bogota   colombia      bogota    220.0

In [125]:
age_dict['15']

Age        City         Country              Region   User-ID
1       18.0    stockton             usa          california       2.0
3       17.0       porto        portugal            v.n.gaia       4.0
19      19.0   langhorne             usa        pennsylvania      20.0
23      19.0     cologne         germany                 nrw      24.0
27      24.0    freiburg         germany  baden-wuerttemberg      28.0
...      ...         ...             ...                 ...       ...
278834  18.0     karachi        pakistan               sindh  278835.0
278837  15.0   massillon             usa                ohio  278838.0
278845  23.0     toronto          canada             ontario  278846.0
278848  23.0  georgetown          canada             ontario  278849.0
278852  17.0   stranraer  united kingdom                 n/a  278853.0

[40179 rows x 5 columns]

In [127]:
age_dict['205']

Age          City        Country            Region   User-ID
52471   209.0        frisco            usa             texas   52472.0
52648   212.0     vancouver         canada  british columbia   52649.0
76864   209.0      arrasate  euskal herria          gipuzkoa   76865.0
92857   210.0     melbourne      australia          victoria   92858.0
156353  210.0       hamburg        germany           hamburg  156354.0
168716  207.0     rochester            usa        washington  168717.0
216665  208.0  christchurch    new zealand        canterbury  216666.0
257113  210.0    glen ridge            usa        new jersey  257114.0

### 11. Find the books that appear only in the goodbooks datasets.

In [89]:
goodbooks[~goodbooks.isbn.isin(books.ISBN)]

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
0         1            2767052       2767052   2792775          272   
1         2                  3             3   4640799          491   
2         3              41865         41865   3212258          226   
3         4               2657          2657   3275794          487   
4         5               4671          4671    245494         1356   
..      ...                ...           ...       ...          ...   
94       95               5297          5297   1858012         2303   
95       96           13536860      13536860  18034963          133   
96       97              17245         17245   3165724         2207   
97       98            5060378       5060378   6976108          209   
98       99           11857408      11857408  16813814          147   

          isbn        isbn13                                    authors  \
0    439023483  9.780439e+12                            Suzanne Collins   
1    439554934  9.780440e+12               J.K. Rowling, Mary GrandPré   
2    316015849  9.780316e+12                            Stephenie Meyer   
3     61120081  9.780061e+12                                 Harper Lee   
4    743273567  9.780743e+12                        F. Scott Fitzgerald   
..         ...           ...                                        ...   
94   375751513  9.780376e+12             Oscar Wilde, Jeffrey Eugenides   
95   345803507  9.780346e+12                                 E.L. James   
96   393970124  9.780394e+12  Bram Stoker, Nina Auerbach, David J. Skal   
97   307269981  9.780307e+12                 Stieg Larsson, Reg Keeland   
98  1612130585  9.781612e+12                                 E.L. James   

    original_publication_year                            original_title  ...  \
0                      2008.0                          The Hunger Games  ...   
1                      1997.0  Harry Potter and the Philosopher's Stone  ...   
2                      2005.0                                  Twilight  ...   
3                      1960.0                     To Kill a Mockingbird  ...   
4                      1925.0                          The Great Gatsby  ...   
..                        ...                                       ...  ...   
94                     1891.0                The Picture of Dorian Gray  ...   
95                     2012.0                        Fifty Shades Freed  ...   
96                     1897.0                                   Dracula  ...   
97                     2006.0               Flickan som lekte med elden  ...   
98                     2011.0                       Fifty Shades Darker  ...   

   ratings_count work_ratings_count  work_text_reviews_count  ratings_1  \
0        4780653            4942365                   155254      66715   
1        4602479            4800065                    75867      75504   
2        3866839            3916824                    95009     456191   
3        3198671            3340896                    72586      60427   
4        2683664            2773745                    51992      86236   
..           ...                ...                      ...        ...   
94        590014             674551                    19647       9825   
95        387290             610116                    25287      39079   
96        618973             689739                    17263      11756   
97        563994             649253                    31423       4623   
98        552059             623340                    28052      37245   

    ratings_2  ratings_3  ratings_4  ratings_5  \
0      127936     560092    1481305    2706317   
1      101676     455024    1156318    3011543   
2      436802     793319     875073    1355439   
3      117415     446835    1001952    1714267   
4      197621     606158     936012     947718   
..        ...        ...        ...        ...   
94      30735     129420     246892   

### 12. Assuming that each pair (author, title) identifies a book, for each book find the number of times it appears in the books dataset. Which books appear the most times?

In [54]:
most_times = books
most_times['Autori'] = most_times['Book-Author'].str.split(',')
most_times = most_times.explode('Autori')
most_times = most_times.groupby(['Book-Title', 'Autori'], as_index=False).count()[['Book-Title', 'Autori', 'ISBN']]
most_times.columns = ['Title', 'Author', 'Num']

In [55]:
most_times.loc[[most_times['Num'].idxmax()]]

Title             Author  Num
114501  Little Women  Louisa May Alcott   21

### 13. Find the author with the highest average rating according to the goodbooks datasets.

In [125]:
gb_media['Autori'] = gb_media['Authors'].str.split(',')
gb_media = gb_media.explode('Autori')[['AVG Rating', 'Autori']]

<ipython-input-125-15fb1862919c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  gb_media['Autori'] = gb_media['Authors'].str.split(',')


In [140]:
gb_authors_avg = gb_media.groupby('Autori', as_index=False).mean()

In [150]:
gb_authors_avg.loc[gb_authors_avg['AVG Rating'].idxmax()]

Autori         Rufus Beck
AVG Rating       4.527866
Name: 32, dtype: object